src/character_attire.cpp

```diff
@@ -1945,6 +1950,171 @@ void outfit::absorb_damage( Character &guy, damage_unit &elem, bodypart_id bp,
+    }
+    damage_unit damage = damage_unit( sp.get_dmg_type(), static_cast<float>( sp.damage( caster ) ),
+                                      0.0f );
+    const bool damage_target = sp.has_flag( spell_flag::LIQUID_DAMAGE_TARGET );
+    const bool damage_armor = sp.has_flag( spell_flag::LIQUID_DAMAGE_ARMOR );
+    bool ignite = sp.has_flag( spell_flag::IGNITE_FLAMMABLE );
+    const bool permanent = sp.has_flag( spell_flag::PERMANENT );
+    int intensity = sp.effect_intensity( caster );
+    int liquid_amount = sp.liquid_volume( caster );
+    const int dur_moves = sp.duration( caster );
+    const efftype_id spell_effect = efftype_id( sp.effect_data() );
+    const time_duration dur_td = time_duration::from_moves( dur_moves );
+    const std::string liquid_name = sp.field()->obj().substance_name;
+    std::list<item> worn_remains;
+    // Liquid will splash on the outermost items first.
+    int liquid_remaining = liquid_amount;
+    for( auto iter = worn.rbegin(); iter != worn.rend(); ) {
+        item &armor = *iter;
+        if( !armor.covers( bp ) || armor.has_flag( flag_INTEGRATED ) ) {
+            ++iter;
+            continue;
+        }
+        const std::string pre_damage_name = armor.tname();
+        if( rng( 1, 100 ) <= armor.get_coverage( bp ) && liquid_remaining > 0 ) {
+            // The item has intercepted the splash to protect its wearer,
+            // now we roll to see if it's affected.
+            guy.add_msg_if_player( m_warning, _( "Your %1s is splashed with %2s %3s." ),
+                                   armor.tname(), get_liquid_descriptor( liquid_remaining ), liquid_name );
+            // A droplet of acid or bile are less likely to ruin a shirt than a whole bucket.
+            if( rng( 1, 200 - armor.breathability( bp ) ) < liquid_remaining ) {
+                // Apply filth to the item. Currently hardcoded because we don't have other item
+                // flags that would make sense for this. It gets its own probability roll here
+                // because it can't use armor_absorb's.
+                if( sp.has_flag( spell_flag::MAKE_FILTHY ) && !armor.has_flag( flag_INTEGRATED ) &&
+                    !armor.has_flag( flag_SEMITANGIBLE ) && !armor.has_flag( flag_PERSONAL ) &&
+                    !armor.has_flag( flag_AURA ) &&
+                    rng( 1, 200 - armor.breathability( bp ) ) < liquid_remaining ) {
+                    add_msg_if_player_sees( guy, m_warning, _( "%1s %2s is covered in filth!" ), guy.disp_name( true,
+                                            true ),
+                                            armor.tname() );
+                    armor.set_flag( json_flag_FILTHY );
+                    guy.on_worn_item_soiled( armor );
+                }
+                // If this is an armor-damaging liquid, the damage is relative to fluid_remaining
+                // and the coverage of the item.
+                if( damage.amount >= 1.0f && damage_armor ) {
```

Comment:
This whole block should probably be replaced to calls with item::damage_armor_durability and other related functions, to avoid duplication. (Also because right now it isn't doing things like checking if the armor is unbreakable, etc